tools/ds-identify

```diff
@@ -272,9 +272,12 @@ blkid_export() {
     # call 'blkid -c /dev/null export', set DI_BLKID_EXPORT_OUT
     cached "$DI_BLKID_EXPORT_OUT" && return 0
     local out="" ret=0
-    out=$(blkid -c /dev/null -o export) && DI_BLKID_EXPORT_OUT="$out" || {
+    local devs="/dev/[hsv]d[a-z]* /dev/xvd[a-z]* /dev/sr*"
```

Comment:
We should populate the list of devs in a function; *BSD will have a different list of devices to check that we'd look for on Linux.

This misses /dev/dasd, /dev/cciss, /dev/bcacheN, /dev/md, /dev/dm-* /dev/loop /dev/mapper/
there may be more.  Is there a particular device type we could ignore?  

Also we could blkid in stages? common devs;  fallback on blkid without list of devs if no datasources are found?

Reply:
> We should populate the list of devs in a function; *BSD will have a different list of devices to check that we'd look for on Linux.
> 

It never actually worked on freebsd at all until recently. That landed under
#625, and now this code is only called on linux. see
[read_fs_info](https://github.com/canonical/cloud-init/blob/e10fc140a76a7bccc0c7cbc3aaae5b9a0ba04261/tools/ds-identify#L389)

> This misses /dev/dasd, /dev/cciss, /dev/bcacheN, /dev/md, /dev/dm-* /dev/loop /dev/mapper/
> there may be more. Is there a particular device type we could ignore?

How would you suggest that we ignore by type?

I'd suggest that all of the devices you listed above are explicitly the ones
I'm intending to ignore.  I'll "fix" it to add s390's dasd, but the rest are
almost exactly the devices I'm intending to ignore.

The  most likely case i can think of for a config disk on such a device would
be dm-*, through some remote device mapper device.


> 
> Also we could blkid in stages? common devs; fallback on blkid without list of devs if no datasources are found?

not without re-working a lot of the simplicity of how ds-identify works. 'read_fsinfo' collects the info and ohter parts of code make sense of it.

Reply:
>
> > This misses /dev/dasd, /dev/cciss, /dev/bcacheN, /dev/md, /dev/dm-* /dev/loop /dev/mapper/
> > there may be more. Is there a particular device type we could ignore?
>
> How would you suggest that we ignore by type?


I'm not sure. /dev/block/* has MAJOR:MINOR symlinks which could be used to
filter out certain block MAJOR values;  but I don't think this is going to be
effective.  LVMs show up a dm-X; and if you have 133 LVMs; we have no reason
not to check all of those.

>
> I'd suggest that all of the devices you listed above are explicitly the ones
> I'm intending to ignore. I'll "fix" it to add s390's dasd, but the rest are
> almost exactly the devices I'm intending to ignore.

But we can't just ignore them.  Anyone's datasource could be on one of these
devices.  This change breaks them.

Reply:
> > > This misses /dev/dasd, /dev/cciss, /dev/bcacheN, /dev/md, /dev/dm-* /dev/loop /dev/mapper/
> > > there may be more. Is there a particular device type we could ignore?
> > 
> > 
> > How would you suggest that we ignore by type?
> 
> I'm not sure. /dev/block/* has MAJOR:MINOR symlinks which could be used to
> filter out certain block MAJOR values; but I don't think this is going to be
> effective.

names  are just as good (or better) of a filter than major/minor values.  names are consistent and human-meaningful.

>  LVMs show up a dm-X; and if you have 133 LVMs; we have no reason
> not to check all of those.

Yes, we do.   There are no clouds that do that.  the list of datasources that read information from blkid are:

 * NoCloud
 * ConfigDrive
 * OpenNebula
 * RbxCloud
 * OVF
 * IBMCloud

If none of those present data on md or lvm devices... then we don't need to check those devices.  No public cloud is going to do that.  That'd require that guests had support for lvm or md, and probably was linux.  The only questionable one in my head is NoCloud and ConfigDrive, and I really doubt anyone does that.

To be concise, we do not have to support, nor *should* we support, all possible combinations of things.  That just results in impossible to maintain matrices.
> 


> > I'd suggest that all of the devices you listed above are explicitly the ones
> > I'm intending to ignore. I'll "fix" it to add s390's dasd, but the rest are
> > almost exactly the devices I'm intending to ignore.
> 
> But we can't just ignore them. Anyone's datasource could be on one of these
> devices. This change breaks them.

Sure we can.  Do you really think that there is a cloud that presents a disk with config information on lvm or md ?

Reply:
>
> > LVMs show up a dm-X; and if you have 133 LVMs; we have no reason
> > not to check all of those.
>
> Yes, we do. There are no clouds that do that. the list of datasources that read information from blkid are:

*that you know of*

>
>     * NoCloud
>
>     * ConfigDrive
>
>     * OpenNebula
>
>     * RbxCloud
>
>     * OVF
>
>     * IBMCloud
>
>
> If none of those present data on md or lvm devices... then we don't need to
> check those devices. No public cloud is going to do that. That'd require
> that guests had support for lvm or md, and probably was linux. The only
> questionable one in my head is NoCloud and ConfigDrive, and I really doubt
> anyone does that.

The Ubuntu Installer is using NoCloud seed, and the rootfs may be on LVM, or
multipath.

https://github.com/CanonicalLtd/subiquity/blob/20.04.3/subiquity/models/subiquity.py#L206


There may be other users of cloud-init and NoCloud that are not *clouds* but
very much like their datasource being found on whatever block device it is
on.


>
> To be concise, we do not have to support, nor _should_ we support, all
> possible combinations of things. That just results in impossible to maintain
> matrices.

There's nothing to maintain.  This is an *optimization* for systems which have
a very large number of block devices which don't match the pattern in this
patch.  If you happen to have a large number of scsi LUNs, blkid will still
take a longer time.

And the filter excludes things that weren't excluded before.  That sucks for
those who were using it.


It would be nice to avoid calling read_fs_info if the DataSource list didn't
include any Datasources which require read_fs_info.


> >
>
> > > I'd suggest that all of the devices you listed above are explicitly the ones
> > > I'm intending to ignore. I'll "fix" it to add s390's dasd, but the rest are
> > > almost exactly the devices I'm intending to ignore.
> >
> >
> > But we can't just ignore them. Anyone's datasource could be on one of these
> > devices. This change breaks them.
>
> Sure we can. Do you really think that there is a cloud that presents a disk with config information on lvm or md ?

It doesn;'t have to be a cloud.  I don't think this is the right approach
generally.

Reply:
a quick word on FreeBSD and [`geom(8)`](https://man.freebsd.org/geom(8)):

[`geom(4)`](https://man.freebsd.org/geom(4)) ***is*** the abstraction layer that creates labels, partitions and crypto and raid mappers and so forth (unless you use zfs and prefer to create your raid there)

furthermore, the type of output we use with `geom` produces an order of magnitude less text that needs to be parsed than `blkid`, or comparable output formats from `geom` itself, so i think we can for now safely ignore this optimisation in the `geom` path, and come back to it when people using FreeBSD complain it's too slow to run `ds-identify` on their (virtual) hardware

Reply:
Its less about the text than about the probing of block devices.  reading/seeking around random-ish locations on a spinning disk because of lvm is slow.  I suspect we can still go through 300 lines of text pretty quickly.

Reply:
> The Ubuntu Installer is using NoCloud seed, and the rootfs may be on LVM, or
> multipath.
> 
> https://github.com/CanonicalLtd/subiquity/blob/20.04.3/subiquity/models/subiquity.py#L206

That wont be affected.  seed directories are up to the OS to have mounted.  Basically /var/lib/cloud/seed has to be on the root filesystem or ds-identify wont see it.  This does  not change that at all.

blkid block devices and partitions on them (and lvm/md/...) for a label of a config disk.  That is a lable like 'cidata' or 'config-2' or 'CONTEXT' or 'CLOUDMD'.  Those filesystems are *not* going to be on lvm or raid devices.

> There may be other users of cloud-init and NoCloud that are not _clouds_ but
> very much like their datasource being found on whatever block device it is
> on.

And most of those use NoCloud disk or tis /var/lib/cloud/seed.  Those are not going to be affected.